File: dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/AI/Models/OpenAIModel.cs
using Microsoft.Bot.Builder;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using Microsoft.Teams.AI.AI.Prompts;
using Microsoft.Teams.AI.AI.Prompts.Sections;
using Microsoft.Teams.AI.AI.Tokenizers;
using Microsoft.Teams.AI.Exceptions;
using Microsoft.Teams.AI.State;
using Microsoft.Teams.AI.Utilities;
using System.ClientModel.Primitives;
using System.Net;
using System.Text.Json;
using OpenAI;
using OAIChat = OpenAI.Chat;
using Azure.AI.OpenAI;
using static Microsoft.Teams.AI.AI.Prompts.CompletionConfiguration;
using System.ClientModel;
using ServiceVersion = Azure.AI.OpenAI.AzureOpenAIClientOptions.ServiceVersion;
using Azure.AI.OpenAI.Chat;
using OpenAI.Chat;
using Microsoft.Teams.AI.Application;
using System.Reflection;
using System.Runtime.CompilerServices;
using Microsoft.Extensions.Options;

[assembly: InternalsVisibleTo("Microsoft.Teams.AI.Tests")]
#pragma warning disable AOAI001 // Type is for evaluation purposes only and is subject to change or removal in future updates. Suppress this diagnostic to proceed.
namespace Microsoft.Teams.AI.AI.Models
{
    /// <summary>
    /// A `IPromptCompletionModel` for calling OpenAI and Azure OpenAI hosted models.
    /// </summary>
    /// <remarks>
    /// The model has been updated to support calling OpenAI's new o1 family of models. That currently 
    /// comes with a few constraints. These constraints are mostly handled for you but are worth noting:
    ///     
    /// * The o1 models introduce a new `max_completion_tokens` parameter and they've deprecated the
    ///   `max_tokens` parameter. The model will automatically convert the incoming `max_tokens` parameter
    ///   to `max_completion_tokens` for you. But you should be aware that o1 has hidden token usage and costs
    ///   that aren't constrained by the `max_completion_tokens` parameter. This means that you may see an
    ///   increase in token usage and costs when using the o1 models. 
    ///   
    /// * The o1 models do not currently support the sending of system messages which just means that the 
    /// `useSystemMessages` parameter is ignored when calling the o1 models.
    /// 
    /// * The o1 models do not currently support setting the `temperature`, `top_p`, and `presence_penalty` 
    ///   parameters so they will be ignored.
    /// 
    /// * The o1 models do not currently support the use of tools so you will need to use the "monologue" 
    ///   augmentation to call actions.
    /// </remarks>
    public class OpenAIModel : IPromptCompletionStreamingModel
    {
        private readonly BaseOpenAIModelOptions _options;
        private readonly ILogger _logger;

        private readonly OpenAIClient _openAIClient;
        private readonly string _deploymentName;
        private readonly bool _useAzure;
        private static readonly JsonSerializerOptions _serializerOptions = new()
        {
            WriteIndented = true,
            Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping
        };

        private static readonly string _userAgent = "AlphaWave";

        /// <summary>
        /// Events emitted by the model.
        /// </summary>
        public PromptCompletionModelEmitter? Events { get; set; } = new();

        /// <summary>
        /// Initializes a new instance of the <see cref="OpenAIModel"/> class.
        /// </summary>
        /// <param name="options">Options for configuring an `OpenAIModel` to call an OpenAI hosted model.</param>
        /// <param name="loggerFactory">The logger factory instance.</param>
        /// <param name="httpClient">HTTP client.</param>
        public OpenAIModel(OpenAIModelOptions options, ILoggerFactory? loggerFactory = null, HttpClient? httpClient = null)
        {
            Verify.ParamNotNull(options);
            Verify.ParamNotNull(options.DefaultModel, "OpenAIModelOptions.DefaultModel");

            _useAzure = false;
            _logger = loggerFactory == null ? NullLogger.Instance : loggerFactory.CreateLogger<OpenAIModel>();

            options.CompletionType = options.CompletionType ?? CompletionType.Chat;
            options.LogRequests = options.LogRequests ?? false;
            options.RetryPolicy = options.RetryPolicy ?? new List<TimeSpan> { TimeSpan.FromMilliseconds(2000), TimeSpan.FromMilliseconds(5000) };
            options.UseSystemMessages = options.UseSystemMessages ?? false;

            OpenAIClientOptions openAIClientOptions = new()
            {
                RetryPolicy = new SequentialDelayRetryPolicy(options.RetryPolicy, options.RetryPolicy.Count)
            };

            openAIClientOptions.AddPolicy(new AddHeaderRequestPolicy("User-Agent", _userAgent), PipelinePosition.PerCall);
            if (httpClient != null)
            {
                openAIClientOptions.Transport = new HttpClientPipelineTransport(httpClient);
            }
            if (!string.IsNullOrEmpty(options.Organization))
            {
                openAIClientOptions.AddPolicy(new AddHeaderRequestPolicy("OpenAI-Organization", options.Organization!), PipelinePosition.PerCall);
            }
            _openAIClient = new OpenAIClient(new ApiKeyCredential(options.ApiKey), openAIClientOptions);

            _deploymentName = options.DefaultModel;
            _options = options;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="OpenAIModel"/> class.
        /// </summary>
        /// <param name="options">Options for configuring an `OpenAIModel` to call an Azure OpenAI hosted model.</param>
        /// <param name="loggerFactory">The logger factory instance.</param>
        /// <param name="httpClient">HTTP client.</param>
        public OpenAIModel(AzureOpenAIModelOptions options, ILoggerFactory? loggerFactory = null, HttpClient? httpClient = null)
        {
            Verify.ParamNotNull(options);
            Verify.ParamNotNull(options.AzureDefaultDeployment, "AzureOpenAIModelOptions.AzureDefaultDeployment");
            Verify.ParamNotNull(options.AzureEndpoint, "AzureOpenAIModelOptions.AzureEndpoint");
            string apiVersion = options.AzureApiVersion ?? "2024-06-01";
            ServiceVersion? serviceVersion = ConvertStringToServiceVersion(apiVersion);
            if (serviceVersion == null)
            {
                throw new ArgumentException($"Model created with an unsupported API version of `{apiVersion}`.");
            }

            _useAzure = true;
            _logger = loggerFactory == null ? NullLogger.Instance : loggerFactory.CreateLogger<OpenAIModel>();

            options.AzureApiVersion = apiVersion;
            options.CompletionType = options.CompletionType ?? CompletionType.Chat;
            options.LogRequests = options.LogRequests ?? false;
            options.RetryPolicy = options.RetryPolicy ?? new List<TimeSpan> { TimeSpan.FromMilliseconds(2000), TimeSpan.FromMilliseconds(5000) };
            options.UseSystemMessages = options.UseSystemMessages ?? false;

            AzureOpenAIClientOptions azureOpenAIClientOptions = new(serviceVersion.Value)
            {
                RetryPolicy = new SequentialDelayRetryPolicy(options.RetryPolicy, options.RetryPolicy.Count)
            };

            azureOpenAIClientOptions.AddPolicy(new AddHeaderRequestPolicy("User-Agent", _userAgent), PipelinePosition.PerCall);
            if (httpClient != null)
            {
                azureOpenAIClientOptions.Transport = new HttpClientPipelineTransport(httpClient);
            }

            Uri uri = new(options.AzureEndpoint);
            if (options.TokenCredential != null)
            {
                _openAIClient = new AzureOpenAIClient(uri, options.TokenCredential, azureOpenAIClientOptions);
            }
            else if (options.AzureApiKey != null && options.AzureApiKey != string.Empty)
            {
                _openAIClient = new AzureOpenAIClient(uri, new ApiKeyCredential(options.AzureApiKey), azureOpenAIClientOptions);
            }
            else
            {
                throw new ArgumentException("token credential or api key is required.");
            }

            _deploymentName = options.AzureDefaultDeployment;
            _options = options;
        }

        /// <inheritdoc/>
        public async Task<PromptResponse> CompletePromptAsync(ITurnContext turnContext, IMemory memory, IPromptFunctions<List<string>> promptFunctions, ITokenizer tokenizer, PromptTemplate promptTemplate, CancellationToken cancellationToken = default)
        {
            if (_options.CompletionType != CompletionType.Chat)
            {
                throw new TeamsAIException("The legacy completion endpoint has been deprecated, please use the chat completions endpoint instead");
            }

            DateTime startTime = DateTime.UtcNow;
            CompletionConfiguration completion = promptTemplate.Configuration.Completion;
            int maxInputTokens = completion.MaxInputTokens;

            if (this._options.Stream == true && Events != null)
            {
                // Signal start of completion
                BeforeCompletionEventArgs beforeCompletionEventArgs = new(turnContext, memory, promptFunctions, tokenizer, promptTemplate, this._options.Stream ?? false);
                Events.OnBeforeCompletion(beforeCompletionEventArgs);
            }

            // Render prompt
            RenderedPromptSection<List<ChatMessage>> prompt = await promptTemplate.Prompt.RenderAsMessagesAsync(turnContext, memory, promptFunctions, tokenizer, maxInputTokens, cancellationToken);
            if (prompt.TooLong)
            {
                return new PromptResponse
                {
                    Status = PromptResponseStatus.TooLong,
                    Error = new($"The generated chat completion prompt had a length of {prompt.Length} tokens which exceeded the MaxInputTokens of {maxInputTokens}.")
                };
            }

            // Get the model to use.
            string model = promptTemplate.Configuration.Completion.Model ?? _deploymentName;
            bool isO1Model = model.StartsWith("o1-");
            bool useSystemMessages = !isO1Model && _options.UseSystemMessages.GetValueOrDefault(false);
            if (!useSystemMessages && prompt.Output.Count > 0 && prompt.Output[0].Role == ChatRole.System)
            {
                prompt.Output[0].Role = ChatRole.User;
            }

            if (_options.LogRequests!.Value)
            {
                _logger.LogTrace("CHAT PROMPT:");
                _logger.LogTrace(JsonSerializer.Serialize(prompt.Output, _serializerOptions));
            }

            // Map to OpenAI ChatMessage
            IEnumerable<OAIChat.ChatMessage> chatMessages = prompt.Output.Select(chatMessage => chatMessage.ToOpenAIChatMessage());

            ChatCompletionOptions chatCompletionOptions = new()
            {
                Temperature = (float)completion.Temperature,
                TopP = (float)completion.TopP,
                PresencePenalty = (float)completion.PresencePenalty,
                FrequencyPenalty = (float)completion.FrequencyPenalty,
            };

            if (isO1Model)
            {
                chatCompletionOptions.MaxOutputTokenCount = completion.MaxTokens;
                chatCompletionOptions.Temperature = 1;
                chatCompletionOptions.TopP = 1;
                chatCompletionOptions.PresencePenalty = 0;
            } else
            {
                // `MaxOutputTokenCount` is not supported for non-o1 Azure OpenAI models, hence it needs to be set for it to work.
                SetMaxTokens(completion.MaxTokens, chatCompletionOptions);
            }

            // Set tools configurations
            bool isToolsAugmentation = promptTemplate.Configuration.Augmentation.Type == Augmentations.AugmentationType.Tools;
            if (isToolsAugmentation)
            {
                chatCompletionOptions.ToolChoice = completion.GetOpenAIChatToolChoice();
                chatCompletionOptions.AllowParallelToolCalls = completion.ParallelToolCalls;

                if (promptTemplate.Actions.Count > 0)
                {
                    foreach (ChatCompletionAction action in promptTemplate.Actions)
                    {
                        chatCompletionOptions.Tools.Add(action.ToChatTool());
                    }
                }
            }

            // Add Azure chat extension configurations
            IDictionary<string, JsonElement>? additionalData = promptTemplate.Configuration.Completion.AdditionalData;
            if (_useAzure)
            {
                AddAzureChatExtensionConfigurations(chatCompletionOptions, additionalData);
            }

            if (_options.LogRequests!.Value)
            {
                _logger.LogTrace("CHAT COMPLETION CONFIG:");
                _logger.LogTrace(JsonSerializer.Serialize(chatCompletionOptions, _serializerOptions));
            }


            PipelineResponse? rawResponse = null;
            ClientResult<ChatCompletion>? chatCompletionsResponse = null;
            PromptResponse promptResponse = new();
            try
            {
                if (this._options.Stream == true)
                {
                    if (_options.LogRequests!.Value)
                    {
                        // TODO: Colorize
                        _logger.LogTrace("STREAM STARTED:");
                    }

                    // Enumerate the stream chunks
                    ChatMessage message = new(ChatRole.Assistant)
                    {
                        Content = ""
                    };
                    AsyncCollectionResult<StreamingChatCompletionUpdate> streamCompletion = _openAIClient.GetChatClient(_deploymentName).CompleteChatStreamingAsync(chatMessages, chatCompletionOptions, cancellationToken);

                    var toolCallBuilder = new StreamingChatToolCallsBuilder();
                    await foreach (StreamingChatCompletionUpdate delta in streamCompletion)
                    {
                        if (delta.Role != null)
                        {
                            string role = delta.Role.ToString();
                            message.Role = new ChatRole(role);
                        }

                        if (delta.ContentUpdate.Count > 0)
                        {
                            message.Content += delta.ContentUpdate[0].Text;
                        }

                        // Handle tool calls
                        if (isToolsAugmentation && delta.ToolCallUpdates != null && delta.ToolCallUpdates.Count > 0)
                        {
                            foreach (var toolCallUpdate in delta.ToolCallUpdates)
                            {
                                toolCallBuilder.Append(toolCallUpdate);
                            }
                        }

                        ChatMessage currDeltaMessage = new(delta)
                        {
                            ActionCalls = message.ActionCalls // Ensure ActionCalls are included
                        };
                        PromptChunk chunk = new()
                        {
                            delta = currDeltaMessage
                        };

                        ChunkReceivedEventArgs args = new(turnContext, memory, chunk);

                        // Signal chunk received
                        if (_options.LogRequests!.Value)
                        {
                            _logger.LogTrace("CHUNK", delta);
                        }

                        Events!.OnChunkReceived(args);
                    }

                    // Add any tool calls to message
                    var toolCalls = toolCallBuilder.Build();
                    if (toolCalls.Count > 0)
                    {
                        message.ActionCalls = new List<ActionCall>();
                        foreach (var toolCall in toolCalls)
                        {
                            var actionCall = new ActionCall(toolCall);
                            message.ActionCalls.Add(actionCall);
                        }
                    }

                    promptResponse.Message = message;

                    // Log stream completion
                    if (_options.LogRequests!.Value)
                    {
                        _logger.LogTrace("STREAM COMPLETED");
                    }
                }
                else {
                    chatCompletionsResponse = await _openAIClient.GetChatClient(model).CompleteChatAsync(chatMessages, chatCompletionOptions, cancellationToken);
                    rawResponse = chatCompletionsResponse.GetRawResponse();
                    promptResponse.Message = new ChatMessage(chatCompletionsResponse.Value);
                }

                promptResponse.Status = PromptResponseStatus.Success;
            }
            catch (ClientResultException e)
            {
                rawResponse = e.GetRawResponse();
                HttpOperationException httpOperationException = new(e);
                if (httpOperationException.StatusCode == (HttpStatusCode)429)
                {
                    promptResponse.Status = PromptResponseStatus.RateLimited;
                    promptResponse.Error = new("The chat completion API returned a rate limit error.");
                }
                else
                {
                    promptResponse.Status = PromptResponseStatus.Error;
                    promptResponse.Error = new($"The chat completion API returned an error status of {httpOperationException.StatusCode}: {httpOperationException.Message}");
                }
            }

            if (_options.LogRequests!.Value)
            {
                _logger.LogTrace("RESPONSE:");
                _logger.LogTrace($"duration {(DateTime.UtcNow - startTime).TotalMilliseconds} ms");
                if (promptResponse.Status == PromptResponseStatus.Success && chatCompletionsResponse != null)
                {
                    _logger.LogTrace(JsonSerializer.Serialize(chatCompletionsResponse.Value, _serializerOptions));
                }

                if (rawResponse != null)
                {
                    _logger.LogTrace($"status {rawResponse!.Status}");
                    if (promptResponse.Status == PromptResponseStatus.RateLimited)
                    {
                        _logger.LogTrace("HEADERS:");
                        _logger.LogTrace(JsonSerializer.Serialize(rawResponse.Headers, _serializerOptions));
                    }
                }
            }

            // Returns if the unsuccessful response
            if (promptResponse.Status != PromptResponseStatus.Success || (chatCompletionsResponse == null && this._options.Stream == false))
            {
                return promptResponse;
            }

            if (chatCompletionsResponse != null)
            {
                // Process response
                ChatCompletion chatCompletion = chatCompletionsResponse.Value;
                List<ActionCall> actionCalls = new();
                IReadOnlyList<ChatToolCall> toolsCalls = chatCompletion.ToolCalls;
                if (isToolsAugmentation && toolsCalls.Count > 0)
                {
                    foreach (ChatToolCall toolCall in toolsCalls)
                    {
                        actionCalls.Add(new ActionCall(toolCall));
                    }
                }
            }

            if (this._options.Stream == true)
            {
                StreamingResponse? streamer = (StreamingResponse?)memory.GetValue("temp.streamer");

                if (streamer == null)
                {
                    throw new TeamsAIException("The streaming object is empty");
                }

                ResponseReceivedEventArgs responseReceivedEventArgs = new(turnContext, memory, promptResponse, streamer);
                Events!.OnResponseReceived(responseReceivedEventArgs);

                // Let any pending events flush before returning
                await Task.Delay(TimeSpan.FromSeconds(0));
            }


            List<ChatMessage>? inputs = new();
            int lastMessage = prompt.Output.Count - 1;

            // Skips the first message which is the prompt
            if (lastMessage > 0 && prompt.Output[lastMessage].Role != ChatRole.Assistant)
            {
                inputs.Add(prompt.Output.ElementAt(lastMessage));

                // Add remaining parallel tools calls
                if (inputs[0].Role == ChatRole.Tool)
                {
                    int i;
                    for (i = prompt.Output.Count - 1; i >= 0; i--)
                    {
                        if (prompt.Output[i].ActionCalls != null && prompt.Output[i].ActionCalls!.Count > 0)
                        {
                            break;
                        }
                    }
                    int firstMessage = i + 1;
                    inputs = prompt.Output.GetRange(firstMessage, prompt.Output.Count - firstMessage);
                }
            }
            promptResponse.Input = inputs;

            return promptResponse;

        }

        private ServiceVersion? ConvertStringToServiceVersion(string apiVersion)
        {
            return apiVersion switch
            {
                "2024-06-01" => ServiceVersion.V2024_06_01,
                "2024-08-01-preview" => ServiceVersion.V2024_08_01_Preview,
                "2024-10-01-preview" => ServiceVersion.V2024_10_01_Preview,
                _ => null,
            };
        }

        private void AddAzureChatExtensionConfigurations(ChatCompletionOptions options, IDictionary<string, JsonElement>? additionalData)
        {
            if (additionalData == null)
            {
                return;
            }

            if (additionalData != null && additionalData.TryGetValue("data_sources", out JsonElement array))
            {
                List<object> entries = array.Deserialize<List<object>>()!;
                foreach (object item in entries)
                {
                    try
                    {
                        ChatDataSource? dataSource = ModelReaderWriter.Read<ChatDataSource>(BinaryData.FromObjectAsJson(item));
                        if (dataSource != null)
                        {
                            options.AddDataSource(dataSource);
                        }
                    }
                    catch (Exception ex)
                    {
                        throw new ArgumentException($"Invalid azure data source configuration {ex.Message}", ex);
                    }
                }
            }
        }

        internal void SetMaxTokens(int maxTokens, ChatCompletionOptions options)
        {
            MethodInfo setMaxTokens = options.GetType().GetMethod("set__deprecatedMaxTokens", BindingFlags.NonPublic | BindingFlags.Instance);
            setMaxTokens.Invoke(options, new object[] { maxTokens });
        }
    }
}
#pragma warning restore AOAI001 // Type is for evaluation purposes only and is subject to change or removal in future updates. Suppress this diagnostic to proceed.

File: python/packages/ai/teams/ai/models/openai_model.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from __future__ import annotations

import asyncio
import json
from dataclasses import dataclass
from logging import Logger
from typing import Callable, List, Optional, Union, cast

import openai
from botbuilder.core import TurnContext
from openai import NOT_GIVEN, AsyncStream
from openai.types import chat, shared_params
from openai.types.chat.chat_completion_message_tool_call_param import Function

from teams.streaming.prompt_chunk import PromptChunk

from ...state import MemoryBase
from ..prompts.message import ActionCall, ActionFunction, Message, MessageContext
from ..prompts.prompt_functions import PromptFunctions
from ..prompts.prompt_template import PromptTemplate
from ..tokenizers import Tokenizer
from .prompt_completion_model import PromptCompletionModel
from .prompt_completion_model_emitter import PromptCompletionModelEmitter
from .prompt_response import PromptResponse


@dataclass
class OpenAIModelOptions:
    """
    Options for configuring an `OpenAIModel` to call an OpenAI hosted model.
    """

    api_key: str
    "API key to use when calling the OpenAI API."

    default_model: str
    "Default model to use for completions."

    endpoint: Optional[str] = None
    "Optional. Endpoint to use when calling the OpenAI API."

    organization: Optional[str] = None
    "Optional. Organization to use when calling the OpenAI API."

    logger: Optional[Logger] = None
    "Optional. When set the model will log requests"

    stream: bool = False
    "Optional. Whether the model's responses should be streamed back."


@dataclass
class AzureOpenAIModelOptions:
    """
    Options for configuring an `OpenAIModel` to call an Azure OpenAI hosted model.
    """

    default_model: str
    "Default name of the Azure OpenAI deployment (model) to use."

    endpoint: str
    "Deployment endpoint to use."

    api_version: str = "2023-05-15"
    "Optional. Version of the API being called. Defaults to `2023-05-15`."

    api_key: Optional[str] = None
    "API key to use when making requests to Azure OpenAI."

    azure_ad_token_provider: Optional[Callable[..., str]] = None
    """Optional. A function that returns an access token for Microsoft Entra 
    (formerly known as Azure Active Directory), which will be invoked in every request.
    """

    organization: Optional[str] = None
    "Optional. Organization to use when calling the OpenAI API."

    logger: Optional[Logger] = None
    "Optional. When set the model will log requests"

    stream: bool = False
    "Optional. Whether the model's responses should be streamed back."


class OpenAIModel(PromptCompletionModel):
    """
    A `PromptCompletionModel` for calling OpenAI and Azure OpenAI hosted models.

    The model has been updated to support OpenAI's new o1/o3 family of models. That currently
    comes with a few constraints. These constraints are mostly handled for you but are worth noting:

    - The models introduce a new `max_completion_tokens` parameter and they've deprecated
    the `max_tokens` parameter. The model will automatically convert the incoming `max_tokens
    ` parameter to `max_completion_tokens` for you. But you should be aware that it has hidden
    token usage and costs that aren't constrained by the `max_completion_tokens` parameter.
    This means that you may see an increase in token usage and costs when using the models.

    - The models do not currently support the sending of system message so the model will
    map them to user message in this case.

    - The models do not currently support setting the `temperature`, `top_p`, and
    `presence_penalty` parameters so they will be ignored.

    - The models do not currently support the use of tools so you will need to use the
    "monologue" augmentation to call actions.
    """

    _options: Union[OpenAIModelOptions, AzureOpenAIModelOptions]
    _client: openai.AsyncOpenAI

    @property
    def options(self) -> Union[OpenAIModelOptions, AzureOpenAIModelOptions]:
        return self._options

    def __init__(self, options: Union[OpenAIModelOptions, AzureOpenAIModelOptions]) -> None:
        """
        Creates a new `OpenAIModel` instance.

        Args:
            options (OpenAIModelOptions | AzureOpenAIModelOptions): model options.
        """

        self._options = options

        if isinstance(options, OpenAIModelOptions):
            self._client = openai.AsyncOpenAI(
                api_key=options.api_key,
                base_url=options.endpoint,
                organization=options.organization,
                default_headers={"User-Agent": self.user_agent},
            )
        elif isinstance(options, AzureOpenAIModelOptions):
            self._client = openai.AsyncAzureOpenAI(
                api_key=options.api_key,
                api_version=options.api_version,
                azure_ad_token_provider=options.azure_ad_token_provider,
                azure_endpoint=options.endpoint,
                azure_deployment=options.default_model,
                organization=options.organization,
                default_headers={"User-Agent": self.user_agent},
            )
        self.events = PromptCompletionModelEmitter()

    async def complete_prompt(
        self,
        context: TurnContext,
        memory: MemoryBase,
        functions: PromptFunctions,
        tokenizer: Tokenizer,
        template: PromptTemplate,
    ) -> PromptResponse[str]:
        # pylint: disable-msg=too-many-locals
        max_input_tokens = template.config.completion.max_input_tokens

        # Setup tools if enabled
        is_tools_aug = (
            template.config.augmentation
            and template.config.augmentation.augmentation_type == "tools"
        )
        tool_choice = (
            template.config.completion.tool_choice
            if template.config.completion.tool_choice is not None
            else "auto"
        )
        parallel_tool_calls = (
            template.config.completion.parallel_tool_calls
            if template.config.completion.parallel_tool_calls is not None
            else True
        )
        tools: List[chat.ChatCompletionToolParam] = []

        # If tools is enabled, reformat actions to schema
        if is_tools_aug and template.actions:
            for action in template.actions:
                curr_tool = chat.ChatCompletionToolParam(
                    type="function",
                    function=shared_params.FunctionDefinition(
                        name=action.name,
                        description=action.description or "",
                        parameters=action.parameters or {},
                    ),
                )
                tools.append(curr_tool)

        model = (
            template.config.completion.model
            if template.config.completion.model is not None
            else self._options.default_model
        )
        is_thinking_model = model.startswith("o1") or model.startswith("o3")

        if self._options.stream and self.events is not None:
            # Signal start of completion
            self.events.emit_before_completion(
                context=context,
                memory=memory,
                functions=functions,
                tokenizer=tokenizer,
                template=template,
                streaming=True,
            )

        res = await template.prompt.render_as_messages(
            context=context,
            memory=memory,
            functions=functions,
            tokenizer=tokenizer,
            max_tokens=max_input_tokens,
        )

        if res.too_long:
            return PromptResponse[str](
                status="too_long",
                error=f"""
                the generated chat completion prompt had a length of {res.length} tokens
                which exceeded the max_input_tokens of {max_input_tokens}
                """,
            )

        if self._options.logger is not None:
            self._options.logger.debug(f"PROMPT:\n{res.output}")

        messages: List[chat.ChatCompletionMessageParam]
        messages = self._map_messages(res.output, is_thinking_model)

        try:
            extra_body = {}
            if template.config.completion.data_sources is not None:
                extra_body["data_sources"] = template.config.completion.data_sources

            max_tokens = template.config.completion.max_tokens
            completion = await self._client.chat.completions.create(
                messages=messages,
                model=model,
                presence_penalty=(
                    template.config.completion.presence_penalty if not is_thinking_model else 0
                ),
                frequency_penalty=template.config.completion.frequency_penalty,
                top_p=template.config.completion.top_p if not is_thinking_model else 1,
                temperature=template.config.completion.temperature if not is_thinking_model else 1,
                max_tokens=max_tokens if not is_thinking_model else NOT_GIVEN,
                max_completion_tokens=max_tokens if is_thinking_model else NOT_GIVEN,
                tools=tools if len(tools) > 0 else NOT_GIVEN,
                tool_choice=tool_choice if len(tools) > 0 else NOT_GIVEN,
                parallel_tool_calls=parallel_tool_calls if len(tools) > 0 else NOT_GIVEN,
                extra_body=extra_body,
                stream=self._options.stream,
            )

            if self._options.stream:
                # Log start of streaming
                if self._options.logger is not None:
                    self._options.logger.debug("STREAM STARTED:")

                # Enumerate the stream chunks
                message_content = ""
                message: Message[str] = Message(role="assistant", content="")
                completion = cast(AsyncStream[chat.ChatCompletionChunk], completion)

                async for chunk in completion:
                    delta = chunk.choices[0].delta

                    if delta.role:
                        message.role = delta.role

                    if delta.content:
                        message_content += delta.content

                    if is_tools_aug and delta.tool_calls:
                        if not hasattr(message, "action_calls") or message.action_calls is None:
                            message.action_calls = []

                        for tool_call in delta.tool_calls:
                            # Add empty tool call to message if new index
                            # Note that a single tool call can span multiple chunks
                            index = tool_call.index

                            if index >= len(message.action_calls):
                                message.action_calls.append(
                                    ActionCall(
                                        id="",
                                        type="function",
                                        function=ActionFunction(
                                            name="",
                                            arguments="",
                                        ),
                                    )
                                )

                            if tool_call.id:
                                message.action_calls[index].id = tool_call.id

                            if tool_call.type:
                                message.action_calls[index].type = tool_call.type

                            if tool_call.function:
                                if tool_call.function.name:
                                    message.action_calls[
                                        index
                                    ].function.name += tool_call.function.name

                                if tool_call.function.arguments:
                                    message.action_calls[
                                        index
                                    ].function.arguments += tool_call.function.arguments

                    if self._options.logger is not None:
                        self._options.logger.debug(f"CHUNK ${delta}")

                    curr_delta_message = PromptChunk(
                        delta=Message[str](
                            role=str(delta.role),
                            content=delta.content,
                            context=(
                                MessageContext.from_dict(delta.context)
                                if hasattr(delta, "context")
                                else None
                            ),
                        )
                    )

                    if self.events is not None:
                        self.events.emit_chunk_received(context, memory, curr_delta_message)

                message.content = message_content

                # Log stream completion
                if self._options.logger is not None:
                    self._options.logger.debug("STREAM COMPLETED:")

                res_input: Optional[Union[Message, List[Message]]] = None
                last_message = len(res.output) - 1

                # Skips the first message which is the prompt
                if last_message > 0 and res.output[last_message].role != "assistant":
                    res_input = res.output[last_message]

                response = PromptResponse[str](input=res_input, message=message)

                streamer = memory.get("temp.streamer")
                if (self.events is not None) and (streamer is not None):
                    self.events.emit_response_received(context, memory, response, streamer)

                # Let any pending events flush before returning
                await asyncio.sleep(0)
                return response

            completion = cast(chat.ChatCompletion, completion)
            if self._options.logger is not None:
                self._options.logger.debug("COMPLETION:\n%s", completion.model_dump_json())

            # Handle tools flow
            action_calls = []
            tool_calls = completion.choices[0].message.tool_calls

            if is_tools_aug and tool_calls:
                for curr_tool_call in tool_calls:
                    action_calls.append(
                        ActionCall(
                            id=curr_tool_call.id,
                            type=curr_tool_call.type,
                            function=ActionFunction(
                                name=curr_tool_call.function.name,
                                arguments=curr_tool_call.function.arguments,
                            ),
                        )
                    )

            input: Optional[Union[Message, List[Message]]] = None
            last_message = len(res.output) - 1

            # Skips the first message which is the prompt
            if last_message > 0 and res.output[last_message].role != "assistant":
                input = res.output[last_message]

                # Add remaining parallel tool calls
                if input.role == "tool":
                    first_message = len(res.output)
                    for msg in reversed(res.output):
                        if msg.action_calls:
                            break
                        first_message -= 1
                    input = res.output[first_message:]

            return PromptResponse[str](
                input=input,
                message=Message(
                    role=completion.choices[0].message.role,
                    content=completion.choices[0].message.content,
                    action_calls=(action_calls if is_tools_aug and len(action_calls) > 0 else None),
                    context=(
                        MessageContext.from_dict(completion.choices[0].message.context)
                        if hasattr(completion.choices[0].message, "context")
                        else None
                    ),
                ),
            )
        except openai.APIError as err:
            if self._options.logger is not None:
                self._options.logger.error("ERROR:\n%s", json.dumps(err.body))

            return PromptResponse[str](
                status="error",
                error=f"""
                The chat completion API returned an error
                status of {err.code}: {err.message}
                """,
            )

    def _map_messages(self, msgs: List[Message], is_thinking_model: bool):
        output = []
        for msg in msgs:
            param: Union[
                chat.ChatCompletionUserMessageParam,
                chat.ChatCompletionAssistantMessageParam,
                chat.ChatCompletionSystemMessageParam,
                chat.ChatCompletionToolMessageParam,
            ] = chat.ChatCompletionUserMessageParam(
                role="user",
                content=msg.content if msg.content is not None else "",
            )

            if msg.name:
                setattr(param, "name", msg.name)

            if msg.role == "assistant":
                param = chat.ChatCompletionAssistantMessageParam(
                    role="assistant",
                    content=msg.content if msg.content is not None else "",
                )

                tool_call_params: List[chat.ChatCompletionMessageToolCallParam] = []

                if msg.action_calls and len(msg.action_calls) > 0:
                    for tool_call in msg.action_calls:
                        tool_call_params.append(
                            chat.ChatCompletionMessageToolCallParam(
                                id=tool_call.id,
                                function=Function(
                                    name=tool_call.function.name,
                                    arguments=tool_call.function.arguments,
                                ),
                                type=tool_call.type,
                            )
                        )
                    param["content"] = None
                    param["tool_calls"] = tool_call_params

                if msg.name:
                    param["name"] = msg.name

            elif msg.role == "tool":
                param = chat.ChatCompletionToolMessageParam(
                    role="tool",
                    tool_call_id=msg.action_call_id if msg.action_call_id else "",
                    content=msg.content if msg.content else "",
                )
            elif msg.role == "system":
                # o1 and o3 models do not support system messages
                if is_thinking_model:
                    param = chat.ChatCompletionUserMessageParam(
                        role="user",
                        content=msg.content if msg.content is not None else "",
                    )
                else:
                    param = chat.ChatCompletionSystemMessageParam(
                        role="system",
                        content=msg.content if msg.content is not None else "",
                    )

                if msg.name:
                    param["name"] = msg.name

            output.append(param)
        return output
